doc/man/local/secure-logging.7
.TH secure\-logging 7 "03 June 2026" 4.10.2 "The secure\-logging manual page"
.SH NAME
.PP
secure\-logging \-\-\- Provides forward integrity and confidentiality for system logs.
.SH SYNOPSIS
.PP
.RS
.nf
$(slog \-\-key\-file <host key file> \-\-mac\-file <MAC file> \-\-logmode <(direct|base64|enc)> ${RAWMSG})
.fi
.RE
.SH DESCRIPTION
.PP
Secure logging is an extension of syslog\-ng OSE which provides system log forward integrity and confidentiality. It is implemented in form of a module and is configured as a template in the syslog\-ng OSE configuration file.
.PP
The main objective of the secure logging module is to provide tamper evident logging, for example to adequately protect log records of a system and to provide a sensor indicating breach\-attempts. The secure logging module achieves this by the authentical encryption of each log record with an individual cryptographic key used only once and protecting the integrity of the whole log archive by a cryptographic authentication code. Every attempt to tamper with an individual log record or the log archive itself is immediately detected during log archive verification. Due to this, an attacker can no longer tamper with log records without being detected.
.PP
To use the log file created by the secure logging module for analysis, the log file must first be decrypted and its integrity must be verified. This is achieved with a command line utility that is part of the secure logging module and is installed as part of the syslog\-ng OSE package. This utility can be integrated into the import module of existing analysis environments.
.PP
The secure logging environment uses a cryptographic key for encrypting log entries. Individual log entries are encrypted with their own keys which are immediately discarded after successful encryption to provide forward integrity. An efficient algorithm generates the key for the next log entry based the key used for encrypting the previous log entry. The resulting chain of keys preserves forward integrity, for example a potential attacker cannot deduce the previous key from the current key.
.PP
To optimize log verification and analysis, a sequence number is added to each log entry. The sequence number is also added to the encryption key creating a one\-to\-one correspondence of the encryption key with the log entry. In order to prevent truncation attacks or deletion of individual log entries, a message authentication code (MAC) is iteratively applied to the complete log archive and stored in a separate file. It is used upon verification to check the integrity of the whole log archive.
.PP
The follwoing is an example of three short original log messages that are sent to a destination with secure logging enabled:
.PP
.RS
.nf
This is a log message
And here comes another log message
This is a log message with a longer text that is processed without any problems
.fi
.RE
.PP
To identify the status of the secure logging environment, check the sequence counter by querying the key file with the \fB\fCslogkey\fR utility:
.PP
.RS
.nf
user@host:~> slogkey \-\-counter /etc/syslog\-ng/host.key 
counter=3
.fi
.RE
.PP
The numbering of log messages starts at zero, due to this, the counter is set to three after processing three messages.
.PP
The output of the secure logging template for the three example messages is shown below. The sequence number that was prepended to each message can be observed. The colon indicates the end of the sequence number and the start of the original message. As three message were processed, the sequence counter of the key is also three.
.PP
.RS
.nf
AAAAAAAAAAA=:k3/dYpLsgO2tUJKSauo6dycIBzW6OTC3pyA9TP+7AnqFgEojBzgC2rcK4OPfRtr8yg==
AQAAAAAAAAA=:smw0ATISVgN+BYEu5d7OLBE7aQhHpK9Ro4MndmNgSVrqhcmRCBCj6DUnD6ku0Z29CKJ0N6LAJUgByX4Ev+g=
AgAAAAAAAAA=:5UVybnKL1EAbgC4CLfd8HpgurjREf4LEN61/yWHSD2hbXjRD4QmQdtbwguT1chzdItKSQASps9QRIvR5Jd4AHzHfqxI4aRgdUBcNbAq26nwUCg5vPWygjmbtQaxZgCJYkry8slxCigmbTVs=
.fi
.RE
.PP
The following is the output of a successful verification run:
.PP
.RS
.nf
0000000000000000: This is a log message
0000000000000001: And here comes another log message
0000000000000002: This is a log message with a longer text that is processed without any problems
.fi
.RE
.PP
The original log messages have been successfully restored, and the sequence counters are also assigned to the clear text messages. This helps in analyzing problems within a particular log entry. As real log files will contain thousands of entries. The sequence counter helps to identify faulty entries.
.PP
Before the secure logging module can be used as part of an existing syslog\-ng OSE installation, several preparatory activities are necessary.
.SH KEY GENERATION
.PP
To bootstrap the system, an initial key (\fB\fCk0\fR) must be created and installed on the log host before secure logging environment is started for the first time.
.PP
The initially created host key (\fB\fCk0\fR) has its counter set to \fB\fC0\fR marking it as the initial host key. This host key must be kept secret and it cannot be disclosed to third any parties. It is required to successfully decrypt and verify log archives processed by the secure logging environment. As each log entry is encrypted with its own key, a new host key will be created after successful processing of a log entry and replaces the previous key. Therefore, the initial host key needs to be stored in a safe place before starting the secure logging environment, as it is deleted from the log host after processing of the first log entry. The following steps must be done before starting the secure logging environment. Steps 1 and 2 are performed with the slogkey utility. See the slogkey.1 manual page for details on how to generate a master key and to derive a host key from it. Step 3 and 4 depend on the actual deployment in a target environment.
.nr step0 0 1
.RS
.IP \n+[step0]
Create a master key.
.IP \n+[step0]
Derive an initial host key (\fB\fCk0\fR) from the master key (\fB\fCkey1\fR). Store the \fB\fCk0\fR key in a safe location that is outside of the log host.
.IP \n+[step0]
Deploy \fB\fCk0\fR on the log host, where the secure logging module is to be used.
.RE
.SH CONFIGURATION
.PP
Secure logging can be configured by adding corresponding statements to the \fB\fCsyslog\-ng.conf\fR file.
.PP
Secure logging is implemented as a template and is configured accordingly. Apart from the actual template configuration, no other setting is required to activate secure logging. 
.PP
Secure logging is activated by adding the following statement in the configuration file:
.PP
.RS
.nf
template("$(slog \-\-key\-file  <host key file> \-\-mac\-file <MAC file> \-\-logmode <(direct|base64|enc)> ${RAWMSG})\\n");
.fi
.RE
.PP
The purpose of the elements within the statement:
.PP
\fB\fCslog\fR
    The name of the secure logging template function. This name can be also be found by calling syslog\-ng OSE with the \fB\fC\-\-module\-registry\fR arguments and checking the \fB\fCtemplate\-func\fR property of the secure logging module in the corresponding output.
.PP
\fB\fC\-\-key\-file\fR or \fB\fC\-k\fR
    The host key file. \fB\fC<host key file>\fR is the full path of the file storing the host key on the log host. If this arguments is not supplied or does not point to a valid key file, syslog\-ng OSE does not start and a displays an error message.
.PP
\fB\fC\-\-mac\-file\fR or \fB\fC\-m\fR
    The MAC file. \fB\fC<MAC file>\fR is the full path of the MAC file on the log host. The file is automatically created upon the initial start. If the path is not correct, syslog\-ng OSE does not start and a displays a corresponding error message.
.PP
\fB\fC\-\-logmode\fR or \fB\fC\-l\fR
    The logmode. \fB\fC<(direct|base64|enc)>\fR is the way to control how a log message shall be presented - as is in plain text, only Base64 encoded or encrypted and Base64 encoded.
.PP
\fB\fC${RAWMSG}\fR
    \fB\fC${RAWMSG}\fR provides access to the original log message received at the source. This macro is only available if the store\-raw\-message flag was set for the source. Otherwise, an empty string is passed to the secure logging template. If access to the original message is not available, for example, if the source does not support the \fB\fCstore\-raw\-message\fR flag, then the \fB\fC${MSG}\fR macro can also be used. In this case, however, the integrity guarantee provided by secure logging is limited to the content that this macro provides and does not protect the complete original message.
.PP
\fB\fC\\n\fR
    \fB\fC\\n\fR is a line separator and its use is important, as the secure logging template expects log entries to be separated. When detecting a line separator, the log entry is regarded as complete and is encrypted with the current host key. Therefore, only a single line separator is allowed.
.PP
The secure logging template can be combined with any source or destination with the following limitations:
.RS
.IP \(bu 2
Sources must be line\-oriented. Secure logging uses a line separator in order to distinguish between individual log entries. Sources which provide data in a different format, for example, in the form of raw data obtained directly from a database system, cannot currently be used with the secure logging template, as the separation of log entries is not clearly defined for this type of data.
.IP \(bu 2
Only sources for which the store\-raw\-message flag is implemented and set do benefit from the integrity guarantee provided by the secure logging template. Secure logging aims at protecting the integrity of complete log messages including all associated meta\-data, such as timestamps and host names. syslog\-ng OSE parses the log message into its internal format and provide easy access to parts of a message through macros. While this is convenient when rewriting log messages, it is not helpful for secure logging. syslog\-ng OSE provides the store\-raw\-message flag which provides access to a copy of the original log message after parsing. This is the log message processed and protected by the secure logging template. If the source does not support the \fB\fCstore\-raw\-message flag\fR, then the \fB\fC${MSG}\fR macro can also be used. However, in this case the integrity guarantee provided by secure logging is limited to the content that this macro provides.
.IP \(bu 2
Log rotation of any kind cannot be used with destinations using secure logging, because log rotate overwrites or deletes previous log files. This compromises the cryptographic chain of trust of the log entries preve recovery. To efficiently handle log files, the secure logging environment features iterative verification. Using iterative verification, a log file can be verified in steps. For this to work, the log file must first be downloaded from the log host, together with the corresponding host key and MAC file to a verification host. After downloading, the log file can be safely deleted from the log host. Verification is then performed on the verification host using the iterative mode of the slogverify utility.
.RE
.SS EXAMPLE: SECURE LOGGING TEMPLATE ON A FILE DESTINATION
.PP
.RS
.nf
#############################################
# Minimal syslog\-ng.conf file with secure logging enabled. Encrypted log
# entries are logged to a single file called /var/log/messages.slog
#

@version: 4.8
@include "scl.conf"

source s_local {
    system();
    internal();
};

source s_network {
    network(
        transport("udp")
        port(514)
        flags(store\-raw\-message)
    );
};

# Secure logging template definition
template secure_logging { 
    template("$(slog \-\-key\-file /etc/syslog\-ng/host.key \-\-mac\-file /etc/syslog\-ng/mac.dat \-\-logmode direct ${RAWMSG})\\n");
};

# This configures a secure logging destination
destination d_local {
    file("/var/log/messages.slog" template(secure_logging));
};

log {
    source(s_local);

    # This source has the raw message flag set
    source(s_network);

    # This statement activates secure logging for this destination
    destination(d_local);
};
.fi
.RE
.SH LOG VERIFICATION
.PP
To analyze the log file created in a secure logging environment, the log files must be decrypted and their integrity be verified. Verification requires both the initial host key (\fB\fCk0\fR) and the corresponding MAC file and is performed with the slogverify utility. It is not normally performed on the log host where the secure logging environment is producing log data. In a typical deployment, log files would be retrieved from the log host and transferred to a central log collector where verification it performed. As verification requires the use of  \fB\fCk0\fR, it should only be performed in a trusted environment.
.SS NORMAL MODE
.PP
In normal mode, a complete log archive is verified at once. In a typical environment, this would mean retrieving a log file together with is MAC file from a log host and retrieving the corresponding initial key \fB\fCk0\fR from a safe location and supplying them to the slogverify utility. A typical call sqeuence for normal mode is presented in the following example:
.PP
.RS
.nf
slogverify \-\-key\-file host0.key \-\-mac\-file mac.dat /var/log/messages.slog \-\-logmode enc /var/log/verified/messages.slog  /var/log/verified/messages
.fi
.RE
.PP
The purpose of the elements within the statement:
.PP
\fB\fChost0.key\fR
    The initial host key (\fB\fCk0\fR). Supplying \fB\fCk0\fR is enough for decrypting all log entries, as the key derivation algorithm is able to generate the necessary keys for all subsequent log entries based on \fB\fCk0\fR\&.
.PP
\fB\fCmac.dat\fR
    The MAC file from the log host.
.PP
\fB\fCenc\fR
    The logmode (here in this example enc for enrypted).
.PP
\fB\fC/var/log/messages.slog\fR
    The file containing the encrypted log entries as retrieved from a log host.
.PP
\fB\fC/var/log/verified/messages\fR
    The file receiving the plain text log after decryption.
.PP
Log files may become too large and not fit into the system memory. Verification is therefore performed in chunks. Each part of the log file is transferred to an internal buffer on which verification is performed. After the buffer has been processed, the next chunk is fetched. An optional buffer argument can be supplied to the slogverify utility in order to change the default buffer size of 1000 log entries to a number suitable for the system on which the verification is performed, for example:
.PP
.RS
.nf
slogverify \-\-key\-file host.key \-\-mac\-file mac.dat \-\-logmode enc /var/log/messages.slog /var/log/verified/messages 8000
.fi
.RE
.PP
For more information on verification in normal mode, [[slogverify|adm\-man\-slogkey]].
.SS ITERATIVE MODE
.PP
Verification in normal mode may not be suitable for some application scenarios. Many log hosts use log rotation in order to preserve storage space. In log rotation, a threshold for the maximum amount of storage space and the number of generations is defined for different type of log files. When either storage space is exhausted or the number of generations is reached, the oldest log file is overwritten by new incoming log data. This procedure is not possible in secure logging environment, as overwriting, i.e. deleting a log file would break the cryptographic chain that is established between the log entries. This comes as no surprise, as one of the main objectives of secure logging is to protect against deletion of log entries by a potential attacker.
.PP
To enable a procedure similar to log rotation, the secure logging environment features an iterative mode. In iterative mode, log files can be split into different files and each of these files can be verified separately. Take care when performing verification in iterative mode, as each of the different log files needs to be accompanied by a copy of the host key and the MAC files present on the system at the time of retrieval. A typical usage scenario for the iterative mode is presented in the following steplist:
.nr step1 0 1
.RS
.IP \n+[step1]
Define a storage threshold for the secure logging file destination. In this example a 500MB threshhold is defined.
.IP \n+[step1]
Let the secure logging environment produce log data that is written to the destination up until the allocated 500MB is reached.
.IP \n+[step1]
Stop the secure logging environment and retrieve the log file, the host key and the MAC files from the log host.
.IP \n+[step1]
Delete the log file on the log host but leave host key and MAC file untouched.
.IP \n+[step1]
Restart the secure logging environment.
.IP \n+[step1]
Perform verification in iterative mode with the log file, the host key and the MAC that was retrieved.
.RE
.PP
Steps 2\-6 have to repeated each time the log file reaches a size of 50 MB. Assuming that the log file parts are named after the iteration, for example \fB\fClog.1\fR, \fB\fClog.2\fR, \fB\fClog.3\fR, etc. and a similar convention is applied to the host keys and MAC files, a typical call sequence for the validation of a log file part in iterative mode after three iterations looks like this:
.PP
.RS
.nf
slogverify \-\-iterative \-\-prev\-key\-file host.key.2 \-\-prev\-mac\-file mac.dat.2 \-\-mac\-file mac.dat \-\-logmode enc /var/log/messages.slog.3 /var/log/verified/messages.3
.fi
.RE
.PP
The purpose of the elements within the statement:
.PP
\fB\fChost.key.2\fR
.PP
.RS
.nf
The host key from the previous iteration. In this example, this is the second iteration.
.fi
.RE
.PP
\fB\fCmac.dat.2\fR
.PP
.RS
.nf
The MAC file from the previous iteration. In the example, verification is performed during the third iteration, so the MAC file from the second iteration is required.
.fi
.RE
.PP
\fB\fCmac.dat\fR
.PP
.RS
.nf
The current MAC file from the log host.
.fi
.RE
.PP

\fB\fCenc\fR
.PP
.RS
.nf
The current logmode, here encrypted.
.fi
.RE
.PP

modules/secure-logging/secure-logging.c
/*
 * Copyright (c) 2019-2025 Airbus Commercial Aircraft
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * As an additional exemption you are allowed to compile & link against the
 * OpenSSL libraries as published by the OpenSSL project. See the file
 * COPYING for details.
 *
 */

#include "plugin.h"
#include "template/simple-function.h"
#include "cfg.h"
#include "uuid.h"
#include "str-format.h"
#include "plugin-types.h"
#include "compat/openssl_support.h"
#include <openssl/evp.h>

#include <glib.h>
#include <locale.h>
#include <unistd.h>
#include <openssl/rand.h>
#include <sys/mman.h>

// Secure logging declarations
#include "utils_slog.h"
#include "slog.h"

/*
 * $(slog [opts] $RAWMSG)
 *
 * Maintains the state of the secure logging template
 *
 * Returns the current encryption key and MAC for the next
 * round and maintains the counter for the total number of
 * of log entries processed.
 *
 * Note: The flag store-raw-message has to be enabled in the
 *       source configuration in order to use the $RAWMSG macro.
 *
 * Options:
 *      -k        Full path to encryption key file
 *      -m        Full path to MAC file
 */
typedef struct _TFSlogState
{
  TFSimpleFuncState super;

  gchar *keypath;
  gchar *macpath;
  guint64 numberOfLogEntries;
  gboolean badKey;
  enum LogMode logmode;
  guchar key[KEY_LENGTH];
  guchar aggMAC[CMAC_LENGTH];
} TFSlogState;

/*
 * Initialize the secure logging template
 */
static gboolean
tf_slog_prepare(LogTemplateFunction *self, gpointer s, LogTemplate *parent, gint argc, gchar *argv[], GError **error)
{
  gboolean is_verbose = FALSE;

  if (TRUE == is_verbose)
    {
      g_print("ENTER tf_slog_prepare: argc: %d\n", argc);
      for (int i = 0; i < argc; i++)
        {
          g_print("   argv[%d]: %s\n", i, argv[i]);
        }
    }

  // Get key filename and store in internal state
  // generate initial BigMAC file
  TFSlogState *state = (TFSlogState *)s;

  gchar *keypathbuffer = NULL;
  gchar *macpathbuffer = NULL;
  GOptionContext *ctx;
  GOptionGroup *grp;
  if ( (mlock(state->key, KEY_LENGTH) != 0) || (mlock(state->aggMAC, CMAC_LENGTH) != 0) )
    {
      msg_warning(SLOG_WARNING_PREFIX, evt_tag_str("Reason", "Unable to acquire memory lock"));
    }

  state->badKey = FALSE;

  SLogOptions options[] =
  {
    { "key-file", 'k', "Name of the host key file", "FILE", NULL },
    { "mac-file", 'm', "Name of the MAC file", "FILE", NULL },
    { "logmode",  'l', "Log mode (direct|base64|enc) whether log is expected as is (direct) or plain but encoded as base64 or encrypted", "LOGMODE", NULL },
    { 0 }
  };

  GOptionEntry slog_options[] =
  {
    { options[0].longname, options[0].shortname, 0, G_OPTION_ARG_CALLBACK, &validFileNameArg, options[0].description, options[0].type },
    { options[1].longname, options[1].shortname, 0, G_OPTION_ARG_FILENAME, &macpathbuffer, options[1].description, options[1].type },
    { options[2].longname, options[2].shortname, 0, G_OPTION_ARG_CALLBACK, &validLogModeArg, options[2].description, NULL },
    { 0 }
  };

  ctx = g_option_context_new("- Secure logging template");
  grp = g_option_group_new("Basic options", "Basic template options", "basic", &options, NULL);

  g_option_group_add_entries(grp, slog_options);
  g_option_context_set_main_group(ctx, grp);

  GError *argError = NULL;

  if (!g_option_context_parse(ctx, &argc, &argv, &argError))
    {
      if (argError != NULL)
        {
          g_propagate_error (error, argError);
        }

      g_option_context_free(ctx);
      munlock(state->key, KEY_LENGTH);
      munlock(state->aggMAC, CMAC_LENGTH);
      return FALSE;
    }

  if (TRUE == is_verbose)
    {
      g_print("AFTER g_option_context_parse tf_slog_prepare: argc: %d\n", argc);
      for (int i = 0; i < argc; i++)
        {
          g_print("   argv[%d]: %s\n", i, argv[i]);
        }
    }

  if (argc < 2)
    {
      state->badKey = TRUE;
      g_set_error(error, LOG_TEMPLATE_ERROR, LOG_TEMPLATE_ERROR_COMPILE,
                  SLOG_ERROR_PREFIX ": Template parsing failed. Invalid number of arguments.\n" \
                  "Usage:\n" \
                  "$(slog --key-file FILE --mac-file FILE --logmode (direct | base64 | enc) $RAWMSG)\n");
      g_option_context_free(ctx);
      munlock(state->key, KEY_LENGTH);
      munlock(state->aggMAC, CMAC_LENGTH);
      return FALSE;
    }

  keypathbuffer = options[0].arg;

  if (keypathbuffer == NULL)
    {
      state->badKey = TRUE;
      g_set_error(error, LOG_TEMPLATE_ERROR, LOG_TEMPLATE_ERROR_COMPILE,
                  SLOG_ERROR_PREFIX ": Template parsing failed. Invalid or missing key file");
      g_option_context_free(ctx);
      munlock(state->key, KEY_LENGTH);
      munlock(state->aggMAC, CMAC_LENGTH);
      return FALSE;
    }

  if (macpathbuffer == NULL)
    {
      state->badKey = TRUE;
      g_set_error(error, LOG_TEMPLATE_ERROR, LOG_TEMPLATE_ERROR_COMPILE,
                  SLOG_ERROR_PREFIX ": Template parsing failed. Invalid or missing MAC file");
      g_option_context_free(ctx);
      munlock(state->key, KEY_LENGTH);
      munlock(state->aggMAC, CMAC_LENGTH);
      return FALSE;
    }

  //-- LogMode (direct|base64|enc) ---
  if (NULL == options[2].arg)
    {
      //-- logmode not provided. Fall back to classic secure-logging which is the
      //   same as --logmode enc
      msg_info(SLOG_INFO_PREFIX, evt_tag_str("Reason", "Old configuration without logmode: Use default enc"));
      state->logmode = LOGMODE_ENCRYPTED;
    }
  else
    {
      //-- variant logmode provided
      char *str_logmode_arg = g_strndup(options[2].arg, PATH_MAX - 1); //-- limit buffer
      g_free(options[2].arg);
      options[2].arg = NULL;
      state->logmode = convert_str_logmode(str_logmode_arg);
      g_free(str_logmode_arg);
      str_logmode_arg = NULL;
      if (LOGMODE_PLAIN_DIRECT != state->logmode && LOGMODE_PLAIN_BASE64 != state->logmode
          && LOGMODE_ENCRYPTED != state->logmode)
        {
          g_print("ERROR tf_slog_prepare: Unknown logmode: %d\n", (gint)state->logmode);
          msg_error(SLOG_ERROR_PREFIX, evt_tag_str("Reason", "tf_slog_prepare: Wrong logmode!"), evt_tag_long("logmode",
                    state->logmode));
          munlock(state->key, KEY_LENGTH);
          munlock(state->aggMAC, CMAC_LENGTH);
          return FALSE;
        }
    } //-- end of logmode provided

  state->numberOfLogEntries = 0;
  state->keypath = keypathbuffer;
  state->macpath = macpathbuffer;
  if (!tf_simple_func_prepare(self, state, parent, argc, argv, error))
    {
      state->badKey = TRUE;
      g_option_context_free(ctx);
      munlock(state->key, KEY_LENGTH);
      munlock(state->aggMAC, CMAC_LENGTH);
      return FALSE;
    }

  // Done with argument parsing
  g_option_context_free(ctx);

  //------------
  // Provide an initial MAC.
  // This first MAC is also stored as a file which path is given by
  // state->macpath specified in the template. It is referenced
  // and updated as aggregated MAC later. When it does not exist, it will
  // be created here.
  //
  // Start Conditions
  // Case-1: No MAC file is available
  //         Both MAC and MAC0 are created.
  //         Normal use case when nothing has been logged so far.
  //         Error when key has been used (counter>0) or MAC0 is available.
  //
  // Case-2: There is an invalid MAC file found.
  //
  // Case-3: There is a valid MAC file but no MAC0
  //
  // Case-4: There is a valid MAC file and an invalid MAC0
  //
  // Case-5: Expected files are available.
  //         Normal use case when syslog-ng is restarted.
  //         Both MAC and MAC0 and keys are available.
  //         The log file already contains entries and previous
  //         process used above files.

  gboolean is_good_start = TRUE;
  guchar key[KEY_LENGTH];
  guchar mac[CMAC_LENGTH];
  char pathMac0[PATH_MAX]; //-- full path of MAC0 file mac0.dat
  memset(key, 0, KEY_LENGTH);
  memset(mac, 0, CMAC_LENGTH);

  if (FALSE == get_path_mac0(state->macpath, pathMac0, PATH_MAX))
    {
      //-- ERROR, never ever
      msg_error(SLOG_ERROR_PREFIX,
                evt_tag_str("Reason", "Failed to get path of MAC to create MAC0"),
                evt_tag_str("File", state->macpath));
    }

  if (!g_file_test(state->macpath, G_FILE_TEST_IS_REGULAR))
    {
      if (g_file_test(pathMac0, G_FILE_TEST_IS_REGULAR))
        {
          //-- No MAC but there is a MAC0. Invalid state. Report a problem.
          msg_error(SLOG_ERROR_PREFIX,
                    evt_tag_str("Reason", "MAC0 found without MAC"),
                    evt_tag_str("File", pathMac0));
          is_good_start = FALSE;
        }
      else
        {
          //-- Neither MAC nor MAC0 found. Check key usage count.
          guint64 key_counter = 42;
          (void) readKey(state->key, &key_counter, state->keypath);
          if (key_counter > 0)
            {
              msg_error(SLOG_ERROR_PREFIX,
                        evt_tag_str("Reason", "Number of log entries is greater than 0 but no MAC files provided"),
                        evt_tag_long("Count", key_counter));
              is_good_start = FALSE;
            }
        }
      /* Case-1 */
      //-- No MAC file available. Normal case when first run.
      //-- create initial aggregated MAC mac0
      create_initial_mac0(key, mac);
      //-- write aggregated MAC and MAC0 file
      (void) writeAggregatedMAC(state->macpath, mac);
      (void) writeAggregatedMAC(pathMac0, mac);
      msg_info(SLOG_INFO_PREFIX,
               evt_tag_str("Reason", "MAC0 and MAC have been created"),
               evt_tag_str("File", state->macpath));
    }
  else
    {
      if (!readAggregatedMAC(state->macpath, state->aggMAC))
        {
          /* Case-2 */
          //-- invalid MAC file
          //-- create initial aggregated MAC mac0
          create_initial_mac0(key, mac);
          //-- write aggregated MAC and MAC0 file
          //   overwrite both MAC and MAC0 file
          (void) writeAggregatedMAC(state->macpath, mac);
          (void) writeAggregatedMAC(pathMac0, mac);
          msg_error(SLOG_ERROR_PREFIX,
                    evt_tag_str("Reason", "Checksum of MAC is invalid! MAC0 and MAC have been re-created!"),
                    evt_tag_str("File", pathMac0));
          is_good_start = FALSE;
        }
      else
        {
          if (!g_file_test(pathMac0, G_FILE_TEST_IS_REGULAR))
            {
              /* Case-3 */
              //-- valid MAC file available
              //-- no MAC0 file available, use copy of valid MAC file
              (void) writeAggregatedMAC(pathMac0, state->aggMAC);
              msg_error(SLOG_ERROR_PREFIX,
                        evt_tag_str("Reason", "No MAC0 found! Dummy MAC0 is provided!"),
                        evt_tag_str("File", pathMac0));
              is_good_start = FALSE;
            }
          else
            {
              if (!readAggregatedMAC(pathMac0, mac))
                {
                  /* Case-4 */
                  //-- valid MAC file available
                  //-- invalid MAC0 file found, overwrite it, use copy of valid MAC file
                  (void) writeAggregatedMAC(pathMac0, state->aggMAC);
                  msg_error(SLOG_ERROR_PREFIX,
                            evt_tag_str("Reason", "Checksum of MAC0 invalid! Dummy MAC0 is provided!"),
                            evt_tag_str("File", pathMac0));
                  is_good_start = FALSE;
                }
            }
        }
    }

  /* Case-5 */
  // All files are expected to be available. Check their state now.

  //-- MAC0 ---
  if (g_file_test(pathMac0, G_FILE_TEST_IS_REGULAR))
    {
      // Read MAC0
      if (!readAggregatedMAC(pathMac0, mac))
        {
          is_good_start = FALSE;
          msg_error(SLOG_ERROR_PREFIX,
                    evt_tag_str("Reason", "Unable to read MAC0"),
                    evt_tag_str("File", pathMac0));
        }
    }
  else
    {
      is_good_start = FALSE;
    }

  //-- MAC ---
  if (g_file_test(state->macpath, G_FILE_TEST_IS_REGULAR))
    {
      // Read aggregated MAC
      if (!readAggregatedMAC(state->macpath, state->aggMAC))
        {
          is_good_start = FALSE;
          msg_error(SLOG_ERROR_PREFIX,
                    evt_tag_str("Reason", "Unable to read aggregated MAC"),
                    evt_tag_str("File", state->macpath));
          if (state->numberOfLogEntries > 0)
            {
              msg_error(SLOG_ERROR_PREFIX,
                        evt_tag_str("Reason", "Aggregated MAC not found or invalid"),
                        evt_tag_str("File", state->macpath));
            }
        }
    }
  else
    {
      is_good_start = FALSE;
    }

  if (TRUE == is_good_start)
    {
      msg_info(SLOG_INFO_PREFIX, evt_tag_str("Reason", "Template with key and MAC file successfully initialized."));
    }
  else
    {
      msg_error(SLOG_ERROR_PREFIX, evt_tag_str("Reason", "Initialization not successful"));
      msg_warning(SLOG_WARNING_PREFIX, evt_tag_str("Reason",
                                                   "It is the users reponsiblility to delete MAC files and the log file and provide a new host.key in order to restart in good conditionl."));
      msg_warning(SLOG_WARNING_PREFIX, evt_tag_str("Reason", "Current setup does not allow full verification!"));
    }

  //------------

  // Read key
  if (!readKey(state->key, (guint64 *) & (state->numberOfLogEntries), state->keypath))
    {
      state->badKey = TRUE;
      msg_error(SLOG_ERROR_PREFIX,
                evt_tag_str("Reason", "Template parsing failed, key file not found or invalid. Reverting to clear text logging!"));
      munlock(state->key, KEY_LENGTH);
      munlock(state->aggMAC, CMAC_LENGTH);
      return TRUE;
    }

  msg_info(SLOG_INFO_PREFIX, evt_tag_str("Reason", "Key successfully loaded"));
  OPENSSL_cleanse(key, sizeof key);
  OPENSSL_cleanse(mac, sizeof mac);
  munlock(state->key, KEY_LENGTH);
  munlock(state->aggMAC, CMAC_LENGTH);
  return TRUE;
}

/*
 * Create a new encrypted log entry
 */
static void
tf_slog_call(LogTemplateFunction *self, gpointer s, const LogTemplateInvokeArgs *args, GString *result,
             LogMessageValueType *type)
{
  TFSlogState *state = (TFSlogState *) s;

  if (LOGMODE_PLAIN_DIRECT != state->logmode && LOGMODE_PLAIN_BASE64 != state->logmode
      && LOGMODE_ENCRYPTED != state->logmode)
    {
      g_print("ERROR tf_slog_call: Unknown logmode: %d\n", (gint)state->logmode);
      msg_error(SLOG_ERROR_PREFIX, evt_tag_str("Reason", "tf_slog_call: Wrong logmode!"), evt_tag_long("logmode",
                state->logmode));
      return;
    }

  *type = LM_VT_STRING;
  // If we do not have a good key, just forward input
  if (state->badKey == TRUE)
    {
      g_string_append (result, args->argv[0]->str);
      return;
    }

  // Compute authenticated encryption of input
  guchar outputmacdata[CMAC_LENGTH];

  // Empty string received? Parsing error?
  if (args->argv[0]->len == 0)
    {
      msg_error(SLOG_ERROR_PREFIX, evt_tag_str("Reason", "String of length 0 received"));
      GString *errorString = g_string_new(SLOG_ERROR_PREFIX ": String of length 0 received");

      sLogEntry(state->numberOfLogEntries,
                errorString,
                state->key,
                state->aggMAC,
                result,
                outputmacdata,
                G_N_ELEMENTS(outputmacdata),
                state->logmode);

      g_string_free(errorString, TRUE);
    }
  else
    {
      sLogEntry(state->numberOfLogEntries,
                args->argv[0],
                state->key,
                state->aggMAC,
                result,
                outputmacdata,
                G_N_ELEMENTS(outputmacdata),
                state->logmode);
    }

  memcpy(state->aggMAC, outputmacdata, CMAC_LENGTH);
  evolveKey(state->key);
  state->numberOfLogEntries++;

  int res = writeKey(state->key, state->numberOfLogEntries, state->keypath);
  if (res == 0)
    {
      msg_error(SLOG_ERROR_PREFIX, evt_tag_str("Reason", "Cannot write key to file"));
      return;
    }

  res = writeAggregatedMAC(state->macpath, state->aggMAC);
  if (res == 0)
    {
      msg_error(SLOG_ERROR_PREFIX, evt_tag_str("Reason", "Unable to write aggregated MAC"), evt_tag_str("File",
                state->macpath));
      return;
    }
}

// Secure logging free state function
void
tf_slog_func_free_state(gpointer s)
{
  TFSlogState *state = (TFSlogState *) s;

  g_free(state->keypath);
  state->keypath = NULL;

  g_free(state->macpath);
  state->macpath = NULL;

  tf_simple_func_free_state((gpointer)&state->super);
}


/*
 * Declare the template function for secure logging
 */
TEMPLATE_FUNCTION(TFSlogState, tf_slog, tf_slog_prepare, tf_simple_func_eval, tf_slog_call, tf_slog_func_free_state,
                  NULL);

static Plugin secure_logging_plugins[] =
{
  TEMPLATE_FUNCTION_PLUGIN(tf_slog, "slog"),
};

gboolean
secure_logging_module_init(PluginContext *context, CfgArgs *args)
{
  plugin_register(context, secure_logging_plugins, G_N_ELEMENTS(secure_logging_plugins));
  return TRUE;
}

const ModuleInfo module_info =
{
  .canonical_name = "secure_logging",
  .version = SYSLOG_NG_VERSION,
  .description = "The secure logging module provides template functions enabling forward integrity and confidentiality of logs.",
  .core_revision = SYSLOG_NG_SOURCE_REVISION,
  .plugins = secure_logging_plugins,
  .plugins_len = G_N_ELEMENTS(secure_logging_plugins),
};
